test.py

```diff
@@ -0,0 +1,226 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+import unittest
+from inspect import cleandoc
+from expr_checker import *
+
+class Dangerous:
+    pass
+
+class Good:
+    a = "hi"
+    evil = Dangerous()
+
+    _secret_stuff = "1+1=3"
+
+    def tell_me_hi(self):
+        return self.a
+    
+    def __enter__(self):
+        return Good()
+
+    def __exit__(self, exec_type, exec_value, traceback):
+        pass
+
+
+class ReadOnlyObject:
+    def __init__(self, x):
+        self.set_x_value(x)
+
+    def set_x_value(self, x):
+        self.x = x
+
+
+_WHITELIST = {str, int, type(None), Good, type(range(0))}
+_ALLOWED_ATTR = {"a", "x", "tell_me_hi", "set_x_value"}
+
+
+class TestFuncChecker(unittest.TestCase):
+    def test_function_call(self):
+        def abc(a, b, c):
+            return f"A: {a}, B: {b}, C: {c}"
+
+        self.assertEqual(
+            eval(expr_checker("abc('aaa', 'bbb', 'ccc')",
+                 _WHITELIST, _ALLOWED_ATTR, {})),
+            "A: aaa, B: bbb, C: ccc"
+        )
+
+        with self.assertRaises(ValueError):
+            eval(expr_checker("abc(a='aaa', c='ccc', b=Dangerous())",
+                 _WHITELIST, _ALLOWED_ATTR, {}))
+
+    def test_comp_expr(self):
+        self.assertEqual(
+            eval(expr_checker(
+                "[(lambda x: x**2)(n) for n in range(1, 11)]", _WHITELIST, _ALLOWED_ATTR, {})),
+            [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
+        )
+
+    def test_attribute(self):
+        a = Good()
+        eval(expr_checker("a.a", _WHITELIST, _ALLOWED_ATTR, {}))
+        eval(expr_checker("a.tell_me_hi()", _WHITELIST, _ALLOWED_ATTR, {}))
+
+        with self.assertRaises(ValueError):
+            eval(expr_checker("a.evil", _WHITELIST, _ALLOWED_ATTR, {}))
+            
+        with self.assertRaises(ValueError):
+            eval(expr_checker("a.gather_secret()", _WHITELIST, _ALLOWED_ATTR, {}))
+
+    def test_reaonly_type(self):
+        a = ReadOnlyObject(42)
+
+        eval(expr_checker("a.set_x_value(25)",
+             _WHITELIST, _ALLOWED_ATTR, {ReadOnlyObject}))
+        exec(expr_checker("a.x = 99", _WHITELIST,
+             _ALLOWED_ATTR, {ReadOnlyObject}))
+
+        self.assertEqual(a.x, 42)
+
+    def test_method_return(self):
+        self.assertEqual(
+            eval(expr_checker("Good().tell_me_hi()",
+                 _WHITELIST, _ALLOWED_ATTR, {})),
+            "hi"
+        )
+
+        with self.assertRaises(ValueError):
```

Comment:
Here can you check the error message ?
so we are sure that we raise that right error